File: polyglot/DesignPatterns/Singleton/singleton.py
# Singleton design pattern

class SingleObject(object):
    mBuilt = False
    mInstance = None
    # can be called once only
    def __init__(self):
        if SingleObject.mBuilt == True:
            raise Exception("SingleObject: cannot create instance")
        SingleObject.mBuilt = True
    def getInstance():
        if SingleObject.mInstance is None:
            SingleObject.mInstance = SingleObject()
        return SingleObject.mInstance
    def showMessage(self):
        print("The single object is alive and well")
    

object1 = SingleObject.getInstance()
object1.showMessage()

object2 = SingleObject.getInstance()
if(object1 is object2): # id(object1) == id(object2)
    print("The two objects are the same")
        
 

File: polyglot/DesignPatterns/Singleton/Singleton.java
// Singleton design pattern

class SingleObject {
  // private constructor
  private SingleObject(){}
  // static data member pointing to unique instance
  private static SingleObject mInstance = null;
  // get the only object available
  public static SingleObject getInstance(){
    if(mInstance == null){
      mInstance = new SingleObject();
    }
    return mInstance;
  }
  public void showMessage(){
    System.out.println("The single object is alive and well");
  }
 }

public class Singleton {
  public static void main(String[] args){
    // will not compile
    // SingleObject nope = new SingleObject();

    SingleObject object1 = SingleObject.getInstance();
    object1.showMessage();

    SingleObject object2 = SingleObject.getInstance();
    if(object1 == object2){ // '==' compares object reference in Java
      System.out.println("The two objects are the same");
    }

  }
}

